### DISCUSSION TITLE
Collect requirements/idea/features for v2.6.0

### DISCUSSION THREAD

**Discussion by OrlinVasilev** [goharbor/harbor#16214] [Ideas]
Hi Community,

as we approach the release of v2.5 it's time to brainstorm the ideas and requirements for the feature list for 2.6!

Please share your thoughts here so we can all work on proper roadmap and achiving it!

Orlix

**Comment by srvsstieber** [comment]
Hi, i'm a happy Harbor User. In our department we are looking for:

- custom roles & permissions
- helm scanners & linters
- harbor edge instances
- gdpr compliant deletion of OIDC importet users

Hope this helps.

Stefan.

**Reply by sixeela** [reply]
+1 for custom roles & permissions.

In addition, would it be possible to study on having both local users (used as "technicals") and LDAP or OIDC ? Is there anything which prevents having both type of users ?

**Reply by dioguerra** [reply]
+1 on deletion OIDC users

**Reply by geoger** [reply]
+1 for custom roles & permissions.

**Comment by sagikazarmark** [comment]
It would be nice to consolidate authn/authz a little bit more: if you enable OIDC (which I believe every 15+ company would do) you end up with:

- an admin user
- OIDC users
- OIDC groups
- project level robot accounts
- system level robot accounts

The result of this little bit fragmented system is that:
- You can't create users manually (kind of okay)
- You can create groups manually, but you can't add users to them (pushing authorization to the OIDC provider which is not always a good idea)
- Only admins can create system robot accounts
- Project owners can't control system robot account access to their projects
- Only the admin user and system robot accounts can talk to the API?

I think it'd be nice to see some IAM-like structure.

Another use case that it would support better is customer access. Currently it's only possible with system robot accounts created by an admin user which is not terrible, but also not a very good experience at the moment.

I'd be happy to talk more about these topics and even get involved in some of the implementation.

**Reply by mail2sandeepd** [reply]
> Hi Community,
> 
> as we approach the release of v2.5 it's time to brainstorm the ideas and requirements for the feature list for 2.6!
> 
> Please share your thoughts here so we can all work on proper roadmap and achiving it!
> 
> Orlix

Thanks Team Harbor, Im a happy Harbor User.
One quick suggestion on security side we should have option to select images scanner as earlier it was Clair or Trivy, Should have more flexibility on images scanner selection or we can have simple document for integrating more scanner with Harbor.

I struggled a lot for integrating additional scanner, still not able to manage for any additional image scanner.

Also few more suggestion as below 
- Ability to export the image scan results in pdf and excel format. Right now all the vulnerabilities are displayed within the browser in the Harbor UI and there is no good way to get them out of the system for further review. 
- Ability to send the scanned results to an email address. Based on certain severity level some kind of notification can be triggered to administrator or the Developer/Project owner email address with the details for further action

Many Thanks to TEAM HARBOR

**Reply by geoger** [reply]
I agree with @sagikazarmark .  Suppose as an OIDC user,  I have 'Project Admin' role for several projects.But I have to create robot account for each Project that I owned to access them. That is  nonsense.

**Reply by OrlinVasilev** [reply]
if you create one robot account you can attach it to multiple projects 
![image](https://user-images.githubusercontent.com/7236111/173334666-374f5f12-393e-4785-a034-0d253f4f466d.png)
Or I'm not getting your ask right :) thank you!

**Reply by geoger** [reply]
@OrlinVasilev Sure, I know that. But it is system robot account ,which only could be created by system admin. As a project owner of several projects, the person could only  create project robot account, and  have to create each robot account for each project. It is not reasonable.

**Reply by OrlinVasilev** [reply]
There are two type of robot accounts one is called system robot account and the second is project robot account! The system robot account can be assigned to multiple projects, the project robot account is bound to single project.

That's by design since you have the harbor instance "the system" and then you have all the projects within! there is no abstraction which can combine multiple projects into one for example tenant.

Is in practice your request is to implement multi-tenancy on different level within Harbor if I'm getting it right ? 

@Vad1mo @wy65701436 please check the above if it's correct? :)  Thank you

**Reply by geoger** [reply]
@OrlinVasilev ,  "multi-tenancy" yeah, something like that.  but I think may be there should be some kind of robot user connected to a ldap/oidc user. Why we need robot account? Usually because ldap/oidc user would use 2FA authentication. So, it is hard to use ldap/oidc user to do automation. so we need robot account.  So what kind of robot account a normal oidc/ldap user required? usually a robot account with same permission with himself.  I guess it is easy to implement.   its name may be "robot$oidc/ldap + user_name+ xxx" :-)

**Reply by Vad1mo** [reply]
@geoger robot accounts are not designed, for the way you intend to use them. If you want to have an token with the same permission as the user, you should use the user's CLI password.

**Reply by geoger** [reply]
<img width="324" alt="image" src="https://user-images.githubusercontent.com/9883978/173525871-ad217cc7-1af7-418a-9f0c-ce3288b063f5.png">
By the way, system robot account seems have performance issue. it takes much time to show

**Reply by geoger** [reply]
> @geoger robot accounts are not designed, for the way you intend to use them. If you want to have an token with the same permission as the user, you should use the user's CLI password.
As OIDC user, the CLI password would expire after the session timeout.  and it would work again after the OIDC user login again , until the session timeout.

**Reply by geoger** [reply]
@Vad1mo

**Reply by geoger** [reply]
@Vad1mo  So, it is impossible to use OIDC user's CLI password to do automation

**Reply by AlenversFr** [reply]
Hey,

Here is how we use robot accounts with our single production instance of Harbor with multi tenancy :
We have around 250 customers.

When a customer subscribe he gets : 
2 harbor project : 
- dev space -> receive all the snapshots and pre-releases with retention policy
- prod space -> receive all production ready releases (immutable) without retention policy (manual cleaning once a year)

2 system robot accounts : 
- a read only robot used for deployment authorized on the 2 spaces
- a write robot used for CI (push) authorized on the 2 spaces

3 OIDC groups :
- maintainer group
- developper group
- reader group

Developpers can pull push in the dev space --> they generate a cli secret 
Developpers can pull only in the prod space --> they generate a cli secret
All the automation is done with the 2 system robot accounts (CI and CD)

If a developper needs to acces to 2 different customer subscription he has to be authorized in the OIDC groups.
For automation purpose we have really few cases where we need to authorize a system robot account on multi customer spaces -> not standard.

It works quite well for us.
Hope it helps ;-)

**Reply by Vad1mo** [reply]
> As OIDC user, the CLI password would expire after the session timeout.  and it would work again after the OIDC user login again , until the session timeout.

It depends a bit on your OIDC setup. However, this is still a genuine problem and was mentioned a few times already in many open issues. I have we can figure out a solution for that problem.

**Reply by sagikazarmark** [reply]
I didn't think my comment would generate this kind of discussion, but it's great.

My primary issue with the current "IAM" model is that there are too many places to manage users/robot accounts and access (regardless of what they are supposed to do according to the current model).

I'm a huge fan of AWS's IAM model: no matter what kind of identity is assigned to your user, access management is the same. So my original suggestion was to consolidate the different identity models Harbor currently has and bring them closer to each other.

Another thing that came up here (and I agree with) is how OIDC is used. Ideally, after authentication users should not be different from each other, no matter how they authenticated. For example: OIDC users cannot use the API at the moment. (There is a chance that they can as sometimes indicated by Harbor developers, but noone actually provided a step-by-step guide for doing so)

These topics are probably worth discussing on a more technical/design level, because it's hard to describe the issues on an abstract level. But I think it's fair to say that there is room for improvement.

I'm happy to participate in discussions aiming to improve IAM in Harbor.

**Reply by Vad1mo** [reply]
I agree that the current authn/autz possibilities are fragmented. This is because a lot of things were added over time based on immediate requirements. 
In a few places, Harbor is already using [casbin ](casbin.org) which offers an enormous set of possibilities. IMHO a good start would be to expose casbin to the end user.  Allow users to deviate from "our" suggested default as they wish. 

I a second step we should introduce Casbin in other areas of Harbor.

**Reply by sagikazarmark** [reply]
Personally, I'm not a huge fan of Casbin. Had a lot of issues with it in the past. Regardless of that dislike, a higher level redesign would make more sense IMO and then exposing Casbin features through an abstract interface could be a next step.

**Reply by Vad1mo** [reply]
@sagikazarmark, what do you consider a better alternative to Casbin in the Golang world? The only ohter thing that comes into my mind would be OPA. However, it has its own merits, especially in the context we are in.

**Reply by sagikazarmark** [reply]
@Vad1mo OPA would be a good example, but my point is I wouldn't expose Casbin directly to users, I'd make it an implementation detail hidden behind a custom interface.

That way introducing OPA at some point would become easier.

**Reply by Vad1mo** [reply]
I understand. Exposing the languages of eg. Casbin or OPA to the user has the enormous advantage that we don't have to reinvent the wheel. No forcing the user to learn our concepts and syntax, no extra documentation needed.

**Comment by losadacris71** [comment]
> Hi Community,
> 
> as we approach the release of v2.5 it's time to brainstorm the ideas and requirements for the feature list for 2.6!
> 
> Please share your thoughts here so we can all work on proper roadmap and achiving it!
> 
> Orlix

Please work with docker and trivy adapter  actual script instalation don't work  --with-trivy 
thanks

**Reply by OrlinVasilev** [reply]
Hi @losadacris71, do you have an open issue for that? I just tested it again and works fine with the offline installer!
If that persist on your side please open [issue](https://github.com/goharbor/harbor/issues)

**Comment by danielzhanghl** [comment]
hi harbor team,
we are looking for redis over TLS to secure the internal pod traffic, as in below ticket.
https://github.com/goharbor/harbor/issues/13223

thanks.

**Reply by doublest** [reply]
+1 for secure redis traffic

**Comment by AlenversFr** [comment]
Hi,
We'd love to see this feature "Add a Security Dashboard"
[https://github.com/goharbor/harbor/issues/16121](url)

Basicaly a simple view at project level with a resume of :
- top common CVE in latest image:tag (image push during last 7days in a project)
- list of latest images:tag (push during last 7 days) with the detail of all critical and Major cve
- A global tendency on the security with a comparison between images pushed a month ago versus now (not sure how to calculate this... cvss score ?)
- the possibility to filter the time period (default 7 days), the level of CVE (Critical by default)

And new API endpoints to GET data : 
- Get all image:tags with a specific CVE as parameter
- Get all CVE on a specific image:tag as parameter
- Get all CVE on all image:tag with a specific project as parameter (group by CVE)
- Get all CVE on all image:tag on all project (full list) with the possibility to filter on the level of CVE

**Comment by dioguerra** [comment]
It would be nice for Harbor to fully suport regex.
There are some use cases like registry replication, where i want to be more selective on the image tags I replicate

**Comment by nyarly** [comment]
The single feature that's preventing us from adopting Harbor is https://github.com/goharbor/harbor/issues/8649 - we'd love to see Cockroach as a backend database, and other cloud-native DBs (e.g. Yugabyte) would be great options for Harbor.

Likewise, I haven't seen confirmation here yet, but if Harbor can use Rook Rados object stores (which, like Cockroach) are S3 compatible, that would be ideal for us.

**Comment by OrlinVasilev** [comment]
@qnetter - hope you can find interesting topics here! :)

**Reply by qnetter** [reply]
Excellent!

**Comment by chrisminton** [comment]
I'd _really_ like to see AWS IAM service accounts able to be used for s3 backends as per https://github.com/goharbor/harbor/issues/16490. This is the sole feature that is preventing us going to production with Harbor 🙏

**Reply by OrlinVasilev** [reply]
pinged the maintainers under the issue!Hope they can help!
is the documentation clear enough for this one?

**Reply by chrisminton** [reply]
I believe so @OrlinVasilev!

**Reply by sagikazarmark** [reply]
I believe there is still a problem on the Docker registry side that keeps us from using IAM service accounts: https://github.com/distribution/distribution/issues/3275

**Reply by Vad1mo** [reply]
This feature depends on upstream docker distribution. 

Given the pace and rate of innovation in docker distribution it makes little sense having this in Harbor present before upstream.

**Reply by OrlinVasilev** [reply]
@tianon any ideas here? :)

**Reply by wy65701436** [reply]
Distribution supports this only in main, not in any release. Harbor has to wait for the distribution v3.0 release off main branch, then bump up.

**Reply by chrisminton** [reply]
sad times. Thanks all for the update 👍

**Comment by geoger** [comment]
I suggest to add such a feature.   As system admin, projects , and members who could access one projects could easily be known by UI.  But if the administrator want to know which projects could be accessed by a specific user ,  it is a little hard. no UI place could help

**Comment by robdewit** [comment]
Using the docker image, I would like to be able to choose English language in the UI with a sane date format - not MM/DD/YY, but something like YYYY-MM-DD or DD/MM/YYYY. I guess uk_GB would do.

We are using the docker-compose environment and I know I could rebuild the whole thing following the documentation with another locale, but having English with an internationally readable date format would really be a nice addition to the docker image.

**Reply by robdewit** [reply]
Oh, OK so it looks like https://github.com/goharbor/harbor/pull/16796 is fixing this. I can't wait!

**Comment by PengWuLin** [comment]
https://github.com/goharbor/harbor/issues/17133 ，Can we have a minimal configuration。I just need docker registry。

**Reply by qnetter** [reply]
> #17133 ，Can we have a minimal configuration。I just need docker registry。

I may be missing something but: if you just need a Docker registry, why not install Docker-distribution instead of Harbor?

**Reply by PengWuLin** [reply]
Thank you for your suggest !
In fact, I should describe it this way, We still need some capabilities of harbor, such as trivy. With the development of business, we will also use other components in the future. Therefore, in the first version, I want to use harbor and add the components I want in subsequent versions.

**Comment by qnetter** [comment]
Thanks - being able to configure only the services you want to use - that’s a clear request.

**Comment by OrlinVasilev** [comment]
Since we have released Harbor 2.6 will lock that discussion for historical reference!
